### ISSUE TITLE
Android app logs only first Radar.on('location') event with react-native app

### ISSUE THREAD

**Issue by vadimshvetsov** [radarlabs/react-native-radar#75]
**Describe the bug**
At this time I can't send location to backend service after the first time. With same code iOS logging more stops but even with iOS logging works not clear. Some points of stops is not logging, some points logged while moving, some points logging properly.

It's to hard to follow the docs, when they abstract the way we can use Location from Google Play Services to the docs. I can't understand why I need to make Firebase app at google account for getting `google-services.json` file for using Location which are not based on Firebase.

**To Reproduce**
I'll try to add here everything I did with native part of Android part.

1. I'd started with installing `react-native-radar` and launching `react-native link react-native-radar`
2. Then make some changes in files (some of them are the result of `react-native-link`

**android/app/build.gradle**
```
dependencies {
      ...
+    implementation project(':react-native-radar')
+    implementation 'io.radar:sdk:2.1.+'
      ...
}
...
+// For working with 28 version incompatibility https://stackoverflow.com/a/49350617/6717050
+configurations.all {
+   resolutionStrategy {
+        force 'com.android.support:support-v4:27.1.0'
+    }
+} 
```
**android/app/src/main/java/com/myapp/MainApplication.java**
```
...
+ import io.radar.sdk.Radar;
+ import io.radar.react.RNRadarPackage;

public class MainApplication extends MultiDexApplication implements ReactApplication {
  private final ReactNativeHost mReactNativeHost = new ReactNativeHost(this) {
    ...
    @Override
    protected List<ReactPackage> getPackages() {
      return Arrays.<ReactPackage>asList(
        new MainReactPackage(),
+      new RNRadarPackage(),
    }
    ...
    @Override
    public void onCreate() {
      super.onCreate();
+    Radar.initialize(< RADAR KEY FOR SURE WITHOUT MISTAKE >);
    }
  }
}
```
**android/settings.gradle**
```
+include ':react-native-radar'
+project(':react-native-radar').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-radar/android')
```
**android/app/src/main/AndroidManifest.xml**
```
    ...
    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
    ...
```

React-Native JS Code:
**app.js**
```
const store = configureStore(); // Create redux store

Radar.on('location', ({ location }) => {
    store.dispatch(sendGeoTrackingData(location)); // This action send location data to server, it works
});
```
**`Start.js` component** - Main screen after user logged in the app. On logout `Radar.stopTracking()` executes.
```
...
const startTrackingWithCheck = async () => {
  const status = await Radar.getPermissionsStatus();
  if (status === 'GRANTED') {
    Radar.startTracking();
    clearInterval(startTrackingInterval);
  }
};

const startTracking = async (userId) => {
  Radar.setPlacesProvider('none');
  Radar.setUserId(String(userId));
  const status = await Radar.getPermissionsStatus();
  if (status !== 'GRANTED') {
    Radar.requestPermissions(true);
    startTrackingInterval = setInterval(startTrackingWithCheck, 5000);
  } else {
    Radar.startTracking();
  }
};

class Start extends Component {
...
  componentDidMount() {
    const { userId } = this.props;
    startTracking(userId);
  }
...
}
```

This code works exactly the same as the code (log only first point and it was the last), that I've added below to be more precise with Android steps in [docs](https://radar.io/documentation/sdk#android-configure). 
> Radar uses the Play Services Location library. If you haven't already configured your project for Play Services, follow the instructions [here](https://developers.google.com/android/guides/google-services-plugin).

Okay, lets add Location Google Play Services library (but it works without it too);

**android/build.gradle**
```
    dependencies {
        classpath 'com.android.tools.build:gradle:3.1.4'
+      classpath 'com.google.gms:google-services:4.2.0'
        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
```
**android/app/build.gradle**
Before I've generated `google-services.json` I've created and app in Firebase and in docs they advice to add `firebase-core` to the dependencies :\  Ofcourse I would be happy to remove it, but without Firebase I can't get proper `google-services.json`.
```
    dependencies {
        ...
+     implementation 'com.google.firebase:firebase-core:16.0.1'
+     implementation 'com.google.android.gms:play-services-location:16.0.0' 
        ...
    }
// Very last line
apply plugin: 'com.google.gms.google-services' // It can't works without google-services.json file
```

**Expected behavior**
`Radar.on('location', callback)` logs location after every stop

**Metadata (please complete the following information):**
 - React Native Radar Version: [2.1.7]
 - React Native: [0.57.3]
 - Platform: [Android]
 - Device [Pixel, Emulator with enabled Google Play Services]
 - Device OS Version [Android 9.0 (Pie)]

I would appreciate any help with Radar setup, because my team is on the edge of giving up with attempting to add Radar to RN app.

**Comment by coolbrow** [comment]
Hey there, thanks for the report. 

You certainly do *not* need firebase for Radar to work properly, so feel free to remove that dependency. As you noted, things will work fine without it.

As for the missed locations, it seems you are using the default tracking options, as described here https://github.com/radarlabs/react-native-radar#background-tracking . If you'd like to get more frequent updates of a user's location I'd suggest setting the `sync` parameter to `all` as the default only sends significant locations.

**Comment by vadimshvetsov** [comment]
Thanks for fast response.

> You certainly do **not** need firebase for Radar to work properly, so feel free to remove that dependency. As you noted, things will work fine without it.

Is it possible to get rid of `Firebase app` and `google-services.json` file alongside? Without this file google services can't get builded in app.

**Comment by coolbrow** [comment]
That's correct, you don't need any part of that. We'll update the docs to provide more clarity around this issue. Thanks for bringing it to our attention. 

Let us know if you have any more issues, we're always happy to help.

**Comment by vadimshvetsov** [comment]
@coolbrow The problem isn't in Firebase and the issue is not resolved) The big problem is that my app log only first location. How I can log all stops, not only first? Is it possible without using "all" in "possbileStateChanges"?

I can't to use `classpath 'com.google.gms:google-services:4.2.0'` without using `google-services.js`. But then I wouldn't use Google Play Location library.

**Comment by coolbrow** [comment]
You do not need the google services gradle plugin or library, that means you can remove:
- `google-services.json`
- `apply plugin: 'com.google.gms.google-services'`
-  `classpath 'com.google.gms:google-services:4.2.0'`

You do not need to manually add `'com.google.android.gms:play-services-location:16.0.0'` as a dependency unless you want to use a newer version than bundled the native SDK (`11.6.2`)

As for your location issue, can you describe the problem a little more? Are you moving at least 100 meters before your next stop? Some details about background tracking:

> ... The SDK will wake up while the user is moving (usually every 2-3 minutes), then shut down when the user stops (usually within 3-5 minutes). To save battery, the SDK will not wake up when stopped, and the user must move at least 100 meters from a stop (sometimes more) to wake up the SDK.

Additionally, with `sync: "possibleStateChanges"`, you'll only get moving location updates when the user is current in a place or geofence, but should still receive all stops.

**Comment by vadimshvetsov** [comment]
Thanks for the response and opening issue.

Yep, I have 4 test users on android moving from office to home back and forth for 2 days. And all of them just log first place where they did install the app.

**Comment by coolbrow** [comment]
If these users are on Android Pie (9.0), it's possible the app is being background restricted with app standby buckets, you can read more here: https://developer.android.com/about/versions/pie/power#buckets

So you only see one update ever from these users or just the first stop? Does the Radar dashboard show any updates?

**Comment by vadimshvetsov** [comment]
Yes, `trackOnce` is working and I see this users on my Radar Dashboard.

**Comment by coolbrow** [comment]
Just to confirm, it sounds like just the client-side callbacks that aren't being delivered? It seems like Radar SDK is tracking as expected otherwise?

**Comment by vadimshvetsov** [comment]
Yep, it looks like I doesn't fire 'location' events on which I was subscribed on. But when I test it on real device or emulator  - the first callback is working, but then there is a silence. It looks like SDK doesn't fire location events when location changes

**Comment by coolbrow** [comment]
That actually sounds like correct behavior, as stated above: 
> With `sync: "possibleStateChanges"`, you'll only get moving location updates when the user is currently in a place or geofence, but should still receive all stops.

So only after 3-5 minutes of being in the same location will the SDK register a stopped location, and unless you have geofences set up, you wont see the moving locations register. 

We suggest not using the emulator or simulator to test location tracking as we've seen spotty behavior in the past around stop detection. 

Are you seeing any errors from the `Radar.on('error', callback)` listener?

**Comment by vadimshvetsov** [comment]
> Are you seeing any errors from the `Radar.on('error', callback)` listener?
Yep and I can't log any error :(
> So only after 3-5 minutes of being in the same location will the SDK register a stopped location, and unless you have geofences set up, you wont see the moving locations register.
I didn't disable geofences but I didn't use it with listener.

All my test users with real android devices and they stop near at the office as well as at the home for a way more then 5 minutes. Same React-Native app code works almost as expected with real devices and simulator and log properly the most of stops (but not all actually). I'm not sure, may be SDK can have a bug or my code - I'm stucked on debugging it and find out the reason of this.

**Comment by coolbrow** [comment]
What do you mean you can't log any errors? It would be helpful to know what errors you've been seeing from that callback, it could point to some misconfiguration somewhere. 

> Same React-Native app code works almost as expected with real devices and simulator and log properly the most of stops

What do you mean by this? You're not seeing this issue for everyone, only a few people? Do they have anything in common with their devices, like OS version or manufacturer?

**Comment by vadimshvetsov** [comment]
I mean that I have this listener and there is no errors he can catch ;( 

I can see that issue is only with Android configuration (it can be any mistake in config or maybe bug in SDK, hope that it's not the truth)>

**Comment by coolbrow** [comment]
I'm sorry I'm not sure I understand, so you're not seeing any errors from the error callback at all?

Are you seeing any warnings or errors from the android device itself, using `logcat` (in Android Studio or through command line)? And are you seeing this issue for everyone or just a few devices, you mentioned above that it works as expected in some cases.

**Comment by vadimshvetsov** [comment]
I'll check devices tommorow and be right back, thank you a lot for trying to help! Yes, the error callback never called. I'll try to debug it with logcat and will come back later.

**Comment by coolbrow** [comment]
No problem, it sounds like maybe the callbacks are being registered incorrectly. Do you mind sharing where in the component lifecycle you set up the Radar callbacks?

**Comment by vadimshvetsov** [comment]
Callbacks set inside app.js code, it's not set inside component lifecycle methods, because app from background is not render anything.

**Comment by vadimshvetsov** [comment]
All android devices log location only one time after installing an app. It seems that issue with tracking in JS code. I can see some devices updating their locations but I can't see this points at my backed server. There are two possible problems in this story - SDK doesn't fire location event or listener doesn't execute callback(

**Comment by coolbrow** [comment]
Since you're executing work in the background off of the main thread it's very likely that the OS is killing the app before your work completes. This would explain why the first, foregrounded, location appears but the rest of the callbacks seem to be not executed. 

Are you logging anything to confirm the callback is being called at all, or just looking at your server for confirmation? It may help to save the location locally or log the callback to double check it's being called.

**Comment by vadimshvetsov** [comment]
I've observed just now that function for setting callback are called only first time. When I launch the app the second time - startTracking executes again, but callback not set this time. What is the best place for placing callback while it's not inside component lifecycle?

**Comment by coolbrow** [comment]
You should set up your listeners at the top level of your js bundle so they're set up as soon as the bundle is loaded into memory. The callback does not need to be setup on every app open, just every time the app is created, your behavior sounds correct. 

What I'm asking is about when the `location` callback is actually **called**. How are you confirming that this callback is only called on the first time? Is it just from your server logs? What I would suggest is to add in local logging of this method to double check that behavior. My hunch is that the callback is actually being called but your server request is being killed before it can complete.

**Comment by vadimshvetsov** [comment]
Yep, my callback on location is set in top level file - app.js after creating the redux store and I try to  log everything at this moment.

My callback looks like below at this moment:
**utils/radar.js**
```js
export const setLocationListener = (store) => {
  console.log('LOCATION LISTENER SET');
  Radar.on('location', ({ location }) => {
    console.log('LOCATION SENT', location);
    store.dispatch(sendGeoTrackingData(location));
  });
  console.log('ERROR LISTENER SET');
  Radar.on('error', (err) => {
    console.log('ERROR', err);
  });
};
```
And **app.js**
```jsx
import React from 'react';
import { Provider } from 'react-redux';

import configureStore from 'store/configureStore';
import { setLocationListener } from 'utils/radar';

import RootSwitch from 'navigators/RootSwitch';

export const store = configureStore();
setLocationListener(store);

const App = () => (
  <Provider store={store}>
    <RootSwitch />
  </Provider>
);

export default App;
```

**Comment by coolbrow** [comment]
And no `LOCATION SENT` or `ERROR` logs from your testing, after the first time? 

How big is your app js bundle? It's possible, but less likely, that loading the bundle takes up too much time and the OS kills us prematurely. Although, I wouldn't generally expect this to be the case.

**Comment by vadimshvetsov** [comment]
No, but some times it's not firing even first time. Only when my app goes to the background I have a chance to get first log, but sometimes I have Location Sent event, but I can't findout what I do wrong

**Comment by vadimshvetsov** [comment]
I've tried to use this options:
```
   Radar.startTracking({
      sync: 'all',
      offline: 'replayOff'
    });
```
But it doesn't help too.

**Comment by vadimshvetsov** [comment]
Logcat doesn't show any location or radar reasoned bugs. Is it possible, that Radar doesn't works correctly with logging Android in background?

After many checks it looks like I'm fully synchronized with docs :(

**Comment by coolbrow** [comment]
Thanks for the update, very weird that the SDK seems to behaving but no callbacks received at all. Just to double check - with the same javascript bundle everything works ok on iOS devices, this issue is android only?

**Comment by vadimshvetsov** [comment]
Yes, for sure. Real devices works and iOS simulator works too. It's weird, I can't find the clue, but maybe it in bundle processing and caching in Android and iOS? I can't imagine next steps I can do for make this works

**Comment by coolbrow** [comment]
Do you mind sharing your `AndroidManifest.xml`, it's possible that somehow the receiver is disabled. This is where we declare the receiver, and after manifest merging, the receiver will end up in your manifest as well:
https://github.com/radarlabs/react-native-radar/blob/70754663992c7efdcc88d1f7f38c447c056b754c/android/src/main/AndroidManifest.xml#L5-L12

**Comment by vadimshvetsov** [comment]
Yes, thank you for your time, Russell!

I've just added this to the app and got one location sent as always. I will try to update all devices with this one and test if this make a change for logging.

Is it crucial thing to logging at the bg?

**Comment by coolbrow** [comment]
Oh I think you misunderstood, this should *already* be in your app - it's manifest merged from the `react-native-radar` sdk, the above is just a link to that source. **Do not add it manually.** 

May I see your `AndroidManifest.xml`, its possible something is disabling it, although not likely.

**Comment by coolbrow** [comment]
My hunch is that your network call is being killed by the OS because it's off the main thread while in the background, and the receiver is already marked as finished by then. Have you tried writing that data to disk and uploading it or inspecting it later? How are you collecting device logs to confirm the callback is never called?

**Comment by vadimshvetsov** [comment]
Aw(( That was the hope :)

Okay, so here is my Manifest without receiver that I was added and deleted after your comments.
```xml
     <manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.myapp">

    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
    <uses-permission android:name="android.permission.INSTALL_PACKAGES"/>
    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission android:name="android.permission.SYSTEM_ALERT_WINDOW"/>

    <!-- Notifications support -->
    <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>

    <!-- for Samsung -->
    <uses-permission android:name="com.sec.android.provider.badge.permission.READ"/>
    <uses-permission android:name="com.sec.android.provider.badge.permission.WRITE"/>

    <!--for htc-->
    <uses-permission android:name="com.htc.launcher.permission.READ_SETTINGS"/>
    <uses-permission android:name="com.htc.launcher.permission.UPDATE_SHORTCUT"/>

    <!--for sony-->
    <uses-permission android:name="com.sonyericsson.home.permission.BROADCAST_BADGE"/>
    <uses-permission android:name="com.sonymobile.home.permission.PROVIDER_INSERT_BADGE"/>

    <!--for apex-->
    <uses-permission android:name="com.anddoes.launcher.permission.UPDATE_COUNT"/>

    <!--for solid-->
    <uses-permission android:name="com.majeur.launcher.permission.UPDATE_BADGE"/>

    <!--for huawei-->
    <uses-permission android:name="com.huawei.android.launcher.permission.CHANGE_BADGE"/>
    <uses-permission android:name="com.huawei.android.launcher.permission.READ_SETTINGS"/>
    <uses-permission android:name="com.huawei.android.launcher.permission.WRITE_SETTINGS"/>

    <!--for ZUK-->
    <uses-permission android:name="android.permission.READ_APP_BADGE"/>

    <!--for OPPO-->
    <uses-permission android:name="com.oppo.launcher.permission.READ_SETTINGS"/>
    <uses-permission android:name="com.oppo.launcher.permission.WRITE_SETTINGS"/>

    <uses-sdk />

    <application
      android:name=".MainApplication"
      android:launchMode="singleTop"
      android:allowBackup="true"
      android:hardwareAccelerated="true"
      android:label="@string/app_name"
      android:windowSoftInputMode="adjustPan"
      android:icon="@mipmap/ic_launcher"
      android:roundIcon="@mipmap/ic_launcher_round"
      android:theme="@style/AppTheme">
      <activity
        android:name=".MainActivity"
        android:screenOrientation="portrait"
        android:label="@string/app_name"
        android:configChanges="keyboard|keyboardHidden|orientation|screenSize"
        android:windowSoftInputMode="adjustPan">
        <intent-filter>
            <action android:name="android.intent.action.MAIN" />
            <category android:name="android.intent.category.LAUNCHER" />
        </intent-filter>
      </activity>
    </application>
</manifest>
```

**Comment by coolbrow** [comment]
Also, do you mind sharing your `index.js` (or whatever you've named it), your top-level js file that gets compiled into the native bundle.

**Comment by vadimshvetsov** [comment]
Yes, thanks:
```jsx
import { AppRegistry } from 'react-native';
import App from './src/app';

AppRegistry.registerComponent('MyApp', () => App);
```

**Comment by coolbrow** [comment]
Ah ok! What happens when you move the Radar listener setup **outside** of the `App` component. I mean moving it to right above  this line:
```js-jsx
AppRegistry.registerComponent('MyApp', () => App);
```

Just importing `App` does not mean the code in `app.js` is run, only when the `MyApp` component is loaded will that code run. On iOS, the root component is always loaded in background but on Android, we only load the JS bundle and don't load the root UI component when the app is in background. This would explain why you only see location pings and updates while the view is already loaded (in foreground). 

Let us know if that fixes your issues.

**Comment by vadimshvetsov** [comment]
Okay, I will try to do anything with it, but at this moment my store is not rehydrated. I would try to move code inside this file. Thanks, will be back soon

**Comment by vadimshvetsov** [comment]
At this moment my root index.js looks like below and it seems that nothing changed
```jsx
import { AppRegistry } from 'react-native';
import React from 'react';
import { Provider } from 'react-redux';

import configureStore from 'store/configureStore';
import { setLocationListener } from 'utils/radar';

import RootSwitch from 'navigators/RootSwitch';

const store = configureStore();
setLocationListener(store);

const App = () => (
  <Provider store={store}>
        <RootSwitch />
  </Provider>
);

AppRegistry.registerComponent('MyApp', () => App);
```

**Comment by coolbrow** [comment]
Thanks, that's very odd, will look into this and report back. We're currently trying to recreate the issue on our side.

**Comment by vadimshvetsov** [comment]
Okay. I'm not sure if it would be convinient to dive into our app, but if you mind to debug it there I can show everything via teamviewer or with anything else. Anyway I'm striving to do anything with it on my own too.

**Comment by coolbrow** [comment]
Out of curiosity, how big is your bundled js file? You can (usually) find it in:
`android/app/build/generated/assets/react/debug/index.android.bundle`

**Comment by vadimshvetsov** [comment]
It's 2.2 mb. Apologies, I've missed your first question about bundle size

**Comment by vadimshvetsov** [comment]
I can't get a clue what's going on. Maybe it is possible to debug on lower level?

**Comment by coolbrow** [comment]
No worries, we're trying to diagnose the issue now. Just to confirm, what is your process for collecting device logs to ensure the callback is called? In our testing, the callback is consistently called in the background, but would expect some network requests to be cancelled - could this be what you've been seeing?

**Comment by vadimshvetsov** [comment]
Okay, I'll try to show how my callback works.

Here is code for listener as mentioned above:
```js
export const setLocationListener = (store) => {
  Radar.on('location', ({ location }) => {
    store.dispatch(sendGeoTrackingData(location));
  });
};
```
I pipe store after creating in index.js for possibility to dispatch an action.
We use `redux-saga` and `axios` for sending request to the server and it's works 100% reliable inside the app.

```js
import { call, takeEvery, select } from 'redux-saga/effects';
import Config from 'react-native-config';

import api from 'utils/api';
import moment from 'utils/moment';

export const sendGeoTrackingData = payload => ({ type: 'user/SEND_GEO_TRACKING_DATA', payload });

function* sendGeoTrackingDataWorker({ payload: { latitude, longitude, accuracy } }) {
  const { accessToken } = yield select(({ session }) => session);
  const headers = { 'X-Auth-Token': accessToken };

  try {
    yield call(
      api.post,
      '/api/v1/write-location',
      {
        location: `${latitude}, ${longitude}`,
        accuracy,
        tracedAt: moment().unix(),
        accessToken,
      },
      {
        headers,
        baseURL: Config.GEO_SERVICE_URL,
      },
    );
  } catch (e) {
    console.error(e);
  }
}

export default function* chatsWatcher() {
  yield takeEvery('user/SEND_GEO_TRACKING_DATA', sendGeoTrackingDataWorker);
}
```

**Comment by vadimshvetsov** [comment]
I'll try to get token from `AsyncStorage` directly and test it, maybe reason of this in this thing, but I'm not sure.

**Comment by coolbrow** [comment]
So that is the off-thread work I am referring to that probably get's killed by the OS in the background. Do you have any **local** and **synchronous** logging to check to see if this code is even getting called. Something like a `console.log()` would work, as long as it happens on the calling thread. 

We're trying to find out if the callbacks are not being called or if your work is being cancelled.

**Comment by vadimshvetsov** [comment]
Before executing server request I can't log - `LOCATION SENT` with Logcat, while my listener looks like this while debugging:
```js
export const setLocationListener = (store) => {
  Radar.on('location', ({ location }) => {
    console.log('LOCATION SENT');
    store.dispatch(sendGeoTrackingData(location));
  });
};
```

**Comment by coolbrow** [comment]
Yeah that should show up in logcat, you still only see that once, at most?

**Comment by vadimshvetsov** [comment]
Nope. I just see console log that `LISTENER SET` when it looks like that:
```js
export const setLocationListener = (store) => {
  console.log('LISTENER SET');
  Radar.on('location', ({ location }) => {
    console.log('LOCATION SENT');
    store.dispatch(sendGeoTrackingData(location));
  });
};
```

**Comment by coolbrow** [comment]
Do you see the `LOCATION SENT` logs at all, even when the location upload works?

**Comment by vadimshvetsov** [comment]
When location updates for the first time in app I can see it one time.

**Comment by coolbrow** [comment]
After you walk around for a bit, how do you collect logs to verify its not working?

**Comment by vadimshvetsov** [comment]
I have a server logs with all requests and there is not requests at all except the first ones. So if it would be even wrong request - it will respond with 4xx anyway. But there is not requests after the first one

**Comment by coolbrow** [comment]
Ah ok, I would like to know if there are any **device** logs to see if `LOCATION SENT` is logged while you're walking around. Like I said, I suspect the network request is getting killed by the Android OS, so that's likely why you cant rely on network requests to verify behavior. It would help to check device logs to verify the callback is being called or not. 

If the OS is killing your request, you'll likely need to start your own background process to handle your request as the process is marked as dead as soon as the callback returns.

**Comment by vadimshvetsov** [comment]
I can't see any logs that can say, that callback called

**Comment by coolbrow** [comment]
Ok that's very interesting and unexpected, we'll do some digging and report back.

**Comment by vadimshvetsov** [comment]
At now I've noticed new thing and I don't know why it behaves like that:
When I tap right bottom button for see apps working in background (like chrome on screenshot) when my app works in foreground - I see location symbol on the right corner of the screen and my app logs location if it's changed. 

<img width="497" alt="screenshot 2019-02-24 13 04 54" src="https://user-images.githubusercontent.com/16336572/53297823-22379180-3835-11e9-895b-a688b9020522.png">

But when app is removed from background it can't log location and execute callback. Same thing with real devices. But maybe I'm wrong, I'll try to dive deeper in that.

**Comment by vadimshvetsov** [comment]
I staged all this things to real test devices and they as work the same - I can see updates on Radar Dashboard but callback never executed

**Comment by vadimshvetsov** [comment]
And I've tried to use a webhook as workaround, but it doesn't make a request. After pressing on `Test` button I'm just getting logout from account 🤔 
![screenshot 2019-02-25 at 12 50 49](https://user-images.githubusercontent.com/16336572/53329299-d51ef280-38fc-11e9-856e-e90398fb3aef.png)

**Comment by coolbrow** [comment]
> Do you have clients with RN apps who use background location on Android? Maybe it can't work from JS thread and we need to send request only from Java code?

Yes we do, and requests will work from the JS thread **as long as the android OS does not kill the thread**. I suspect your network request is getting killed because the `BroadcastReceiver` has finished and is marked as complete. You'll need to start up a headless process that doesn't get killed to make network requests or do work off the main thread in the background.

**Comment by vadimshvetsov** [comment]
Do you have any example with Radar and headless JS?

At this moment I’m trying to use webhook but it doesn’t work.
Could you please check why?
I have location updates on Radar internal dashboard but webhook doesn’t send

**Comment by coolbrow** [comment]
Webhooks are for events only, are you seeing any events in your dashboard? What errors are you seeing?

**Comment by vadimshvetsov** [comment]
Oh, okay, we don't use any events, only location updates. Maybe we can use events somehow for that without geofencec?

P.S. I will try to make headless process for callback. I can't find any docs about that, I need to move callback definition to this process?

**Comment by coolbrow** [comment]
The callback definition looks fine, it's the network request that is possibly getting killed by the OS. You'll need to make sure the request and any other off thread work you do is launched in a way that the OS won't kill. On iOS there is a few seconds of grace period before the app is shutdown, this could explain why you're only seeing issues on Android. We're continuing to investigate this issue on our end.

**Comment by vadimshvetsov** [comment]
https://facebook.github.io/react-native/docs/headless-js-android#caveats
If I will define headless JS process it can't be used when app in foreground or app will crash.
Maybe better find a working solution with foreground and background? Could you provide any debugging tips for that?

I will try to remove `redux-saga` and store from the chain and rewrite this without any middleware. Just axios and JS.

**Comment by coolbrow** [comment]
From that link:
> By default, your app will crash if you try to run a task while the app is in the foreground. This is to prevent developers from shooting themselves in the foot by doing a lot of work in a task and slowing the UI. You can pass a fourth boolean argument to control this behaviour.

Sounds like you can just pass 'true' to prevent crashes in the foreground.

**Comment by vadimshvetsov** [comment]
Okay, thanks. At first I'll try request with clean axios, the will look into Headless Android setup.

**Comment by vadimshvetsov** [comment]
I've did plain fetch request and it still working the same, the only way to make it works probably use a Headless JS in RN. Sad(

**Comment by vadimshvetsov** [comment]
Guys, any news on this? I still can't launch that, but team want to finish this thing actually

**Comment by russellcullen** [comment]
Did using a headless task not solve the issue? What other methods have you tried in the meantime?

**Comment by vadimshvetsov** [comment]
I've came back from vacation, I will try to use headless mode, but anyway it must work without it from the box, isn't?

**Comment by russellcullen** [comment]
I'm not sure what you mean - any work that is happening off the main thread and in the background has a potential to be killed by the Android OS. See [this section](https://developer.android.com/guide/components/broadcasts#effects-process-state ) of the docs:
> For this reason, you should not start long running background threads from a broadcast receiver. After onReceive(), the system can kill the process at any time to reclaim memory, and in doing so, it terminates the spawned thread running in the process. 

The Radar SDK does not provide the functionality to run code off-thread and in the background out of the box, this is something that is up to you to implement correctly as it's very use-case dependent. Once the callback returns, any work happening after or concurrently isn't guaranteed to finish, including your network requests. A headless task is just one way to work around this issue, but certainly not the only way.

**Comment by narodejesus** [comment]
Currently trying this library before I make a subscription but
same thing happening when I tried this in RN in android. I read all the comments in here and try to follow it but still only first location is being tracked.

**Comment by nickpatrick** [comment]
Hi, please check the background tracking documentation (https://radar.io/documentation/sdk/react-native#background) and try again on a real device with the latest version. Two things to keep in mind:
- See the tracking options reference (https://radar.io/documentation/sdk/tracking) to understand the behavior of each preset: efficient, responsive, and continuous. Note that `Radar.on('location')` is not called on every location update for efficient and responsive because `sync` = `STOPS_AND_EXITS`. However, `Radar.on('clientLocation')` will be called on every location update regardless of `sync`. Depending on your use case, you may want to use `Radar.on('clientLocation')` instead.
- Make sure that you are adding event listeners outside of your component lifecycle (e.g., outside of componentDidMount) if you want them to work when the app is in the background.
